Run R scripts as asynchronous jobs

jmcphers opened the pull request:
This change brings together two existing pieces of infrastructure -- the Progress API and the `AsyncRProcess` wrapper -- to make it possible to run any R script as a background task. This is useful for performing long or expensive jobs without juggling running multiple copies of RStudio Desktop. 

Any R script can now be sourced as a Job:

![image](https://user-images.githubusercontent.com/470418/41008260-ed4fa6d6-68de-11e8-80ba-0f53d1ce96ea.png)

By default, jobs run in a clean R session and use the working directory of the script being run.

![image](https://user-images.githubusercontent.com/470418/41008316-38520e6c-68df-11e8-8dbc-03f1bd2c6ed6.png)

However it's possible to export the R session's global environment to the job, and it's also possible to export any variables the job creates back into the current R session. (The latter of course should be used with caution since it can overwrite existing values silently!)

Once a job starts, its output is shown in the Jobs tab (which has moved next to the Console):

![image](https://user-images.githubusercontent.com/470418/41008422-ce0e7cf6-68df-11e8-9122-8a85eacce9e3.png)

The Jobs tab shows a summary of current and past jobs:

![image](https://user-images.githubusercontent.com/470418/41008472-06db01e4-68e0-11e8-8901-249c124f5ad5.png)

Note that this feature is not fully complete (this PR just represents the first and largest chunk of work); the primary remaining work items are smoothing out the edges around cancellation, suspension, desktop app exit, launch options, and UX polish.

jjallaire commented:
How about:

**Environment**

- [ ] Run job with copy of current global environment
- [ ] After job, copy objects back to global environment

kevinushey commented:
Does the load of values in the global environment occur automatically after the job is completed? This feels a bit dangerous -- are we sure this should happen automatically rather than through some explicit user gesture? (E.g. click on a completed job in the UI, and explicitly request "Copy generated R objects to global environment" or similar).

I'm imagining users might be running scripts that could create lots of 'temporary' variables in the global environment, so it's possible they could cause a lot of 'pollution' after the copy as well.

Alternatively, would it be worth adding an option to copy the generated values into their own R environment, rather than overlaying directly into the global environment? E.g. an option to run a job whose values would later be copied into an R environment called `job_abc123` in the global environment, which the user could then harvest as needed when the values are available.

What tools around job synchronization will users have? E.g. are we going to have a workflow like:

```
# suppose 'script.R' generates a variable 'abc'
job <- rstudioapi::jobSource("script.R")
status <- rstudioapi::jobWait(job)
print(abc)
```

Similarly, will users have a mechanism for easily doing this kind of 'wait-for-job' for jobs launched through the RStudio UI?

kevinushey commented on src/cpp/session/modules/ModuleTools.R:
```diff
@@ -456,3 +456,8 @@ if (getRversion() < "3.5") {
 }
+
+.rs.addFunction("sessionModulePath", function() {
+   .Call("rs_sessionModulePath")
```
I think we should enforce using `PACKAGE = "(embedding)"` on any new `.Call()` usages, as otherwise you're asking R to crawl through all of the R DLLs to find something called `rs_sessionModulePath` and IIUC the RStudio DLL can be on the tail end of the search list.

kevinushey commented on src/cpp/session/modules/SessionJobs.R:
```diff
@@ -78,3 +78,14 @@
 })
 
+.rs.addApiFunction("runScriptJob", function(path, workingDir = NULL, 
+                                            importEnv = FALSE,
```
Is it intentional that `importEnv` and `exportEnv` are not passed through to the `.Call()` bit below?

kevinushey commented on src/cpp/session/modules/SourceWithProgress.R:
```diff
@@ -0,0 +1,104 @@
+   sourceEnv <- new.env(parent = globalenv())
+
+   # parse the script
+   statements <- parse(file = script, keep.source = TRUE)
```
Do we need to be sensitive to the file encoding here?

kevinushey commented on src/cpp/session/modules/SourceWithProgress.R:
```diff
@@ -0,0 +1,104 @@
+   }
+
+   # find the sections
+   lines <- readLines(script)
```
Similar comment re: encoding?

kevinushey commented on src/cpp/session/modules/jobs/JobsApi.cpp:
```diff
@@ -0,0 +1,184 @@
+namespace {
+
+// map of job ID to jobs
+std::map<std::string, boost::shared_ptr<Job> > s_jobs;
```
Is it okay to have a complex object like this as a static object? I recall we had RStudio crashes-on-exit when attempting to destruct maps like this in the past; it might be better just to explicitly `new` and leak this.

kevinushey replied:
For context: https://github.com/rstudio/rstudio/blob/9a2cd883d53a98b7b6c0957d6fc492e673ee5aac/src/cpp/session/SessionRpc.cpp#L297-L302

jmcphers replied:
That's a good question -- this code just moved from a different file, where it's been in the dailies/previews for a while now without issue. Presumably if this were a problem we'd have seen it by now?

kevinushey replied:
I think you're right -- if we've been running safely with it for a while then I think no change is needed.

kevinushey commented on src/cpp/session/modules/jobs/ScriptJob.cpp:
```diff
@@ -0,0 +1,334 @@
+         // prepare the environment for the script by exporting the current env
+         setJobStatus(job_, "Preparing environment");
+         r::exec::RFunction save("save.image");
+         save.addParam("file", import_.absolutePath());
```
Does this need `utf8ToSystem()`?

jmcphers replied:
No, I don't think we need that when passing parameters directly to R; my understanding is that it's only necessary when preparing system command arguments (we do use it elsewhere in this file for that purpose). Does that align with your understanding?

kevinushey replied:
I think the main issue is that the string is passed to R with `"unknown"` encoding, and so R will assume it's in the system default encoding. Since we're calling `save.image()` directly, I think we need to give it a path in the system-native encoding.

When we invoke our own R functions, we often explicitly re-assign the encoding as UTF-8 to work around this; we won't have that opportunity when calling `save.image()` directly.

jmcphers replied:
Got it, thanks!

kevinushey commented on src/cpp/session/modules/jobs/ScriptJob.cpp:
```diff
@@ -0,0 +1,334 @@
+      {
+         setJobStatus(job_, "Loading results");
+         r::exec::RFunction load("load");
+         load.addParam("file", export_.absolutePath());
```
`utf8ToSystem()`?

kevinushey replied:
I think this call still needs `utf8ToSystem()`

kevinushey commented on src/cpp/session/modules/jobs/SessionJobs.cpp:
```diff
@@ -368,6 +354,7 @@ core::Error initialize()
    RS_REGISTER_CALL_METHOD(rs_setJobStatus, 2);
    RS_REGISTER_CALL_METHOD(rs_setJobState, 2);
    RS_REGISTER_CALL_METHOD(rs_addJobOutput, 3);
+   RS_REGISTER_CALL_METHOD(rs_runScriptJob, 1);
```
Number of arguments here? (Looks like the function accepts 4 arguments)

kevinushey commented on src/gwt/src/org/rstudio/studio/client/workbench/commands/Commands.cmd.xml:
```diff
@@ -3203,17 +3203,32 @@ well as menu structures (for main menu and popup menus).
+
+    <cmd id="sourceAsJob"
+         menuLabel="Source as Job..."
+         desc="Run an the current R script as a job"
```
nit: remove 'an'

jmcphers commented:
> Alternatively, would it be worth adding an option to copy the generated values into their own R environment, rather than overlaying directly into the global environment?

That's a good idea; I've implemented this as an option. There's a tradeoff between simplicity and safety here for sure; I think most of the time users (especially new ones) are going to want to just work with the values right away instead of having to click through an import process or write R code to read them out of an environment. It certainly isn't the safest route, but I think it's the simplest mental model.

> What tools around job synchronization will users have? 

None. :-)

I've added and RStudio API entry to make testing easier, but this whole system is fairly heavily slanted towards interactive use. If users want to invoke R asynchronously *from R* in order to compose a longer workflow, I don't think we should encourage them to use RStudio-specific APIs or mechanisms -- `callr` and friends are a better choice. 

Very much appreciate the thorough review and feedback!

jmcphers commented:
@jjallaire Text updated -- here's the section of the dialog now:

![image](https://user-images.githubusercontent.com/470418/41062289-46da984a-698a-11e8-8504-f94bacfd02e9.png)

(you can also choose to do nothing, or to put the objects into some other environment for safety as @kevinushey suggested.)

kevinushey commented:
LGTM! It's awesome to see the jobs work coming together, this is going to be a huge win for our users :-)

jjallaire commented:
I feel like the default should not be to copy objects into the global environment (as that would amount to an asynchronous overwrite of your current working set). Especially if the job has the semantics of a "fork" which then uses the global environment to do various intermediate calculations it will be very surprising when your current environment suddenly changes to these intermediate values!

I think we might want to change the text to "Copy job results: " with "(Don't copy)" as the default and options  "To the global environment" "To results object in global environment"

That language is actually pretty awkward, but something like that :-)

jmcphers commented:
> I feel like the default should not be to copy objects into the global environment 

This is already the default, for exactly the reasons you mention! 

> I think we might want to change the text to "Copy job results: " with "(Don't copy)" as the default and options "To the global environment" "To results object in global environment"

Thanks, that's an improvement. I'll make this change on my branch.
